pitch pages

kziemianek opened the pull request:
Refs: https://github.com/gitcoinco/web/issues/506


##### Description
Add pitch pages as described #506 

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] linter status: 100% pass
- [x] changes don't break existing behavior
- [x] commit message follows [commit guidelines]

##### Affected core subsystem(s)

- frontend
- backend

##### Testing
Performed manual local tests.
##### Refers/Fixes
Refs #506

kziemianek commented on app/retail/templates/idea_template.html:
```diff
@@ -0,0 +1,51 @@
+<script id="idea_template" type="text/x-jsrender">
```
this file needs to be formated properly ...

kziemianek commented on app/retail/views.py:
```diff
@@ -459,6 +463,88 @@ def browser_extension_chrome(request):
+        github_username=request.POST.get('githubUsername'),
+        summary=request.POST.get('summary'),
+        more_info=request.POST.get('moreInfo'),
+        looking_for_capital=True, #request.POST.get('lookingForCapital') returns 'fasle' or 'true'
```
I have no idea how to serialize properly javascript's true/false to python's True/False, any ideas?

mkosowsk replied:
Per https://stackoverflow.com/questions/10693630/how-to-pass-a-boolean-from-javascript-to-python

mkosowsk replied:
Per https://stackoverflow.com/questions/10693630/how-to-pass-a-boolean-from-javascript-to-python

mkosowsk replied:
Per https://stackoverflow.com/questions/10693630/how-to-pass-a-boolean-from-javascript-to-python

mkosowsk replied:
Per [this](https://stackoverflow.com/questions/10693630/how-to-pass-a-boolean-from-javascript-to-python )

> looking_for_capital = True if request.POST.get('lookingForCapital') == 'true' else False
or
> looking_for_capital = request.POST.get('lookingForCapital') == 'true'

might be a good solution 🤔

mkosowsk replied:
Per [this](https://stackoverflow.com/a/10693668)

> looking_for_capital = True if request.POST.get('lookingForCapital') == 'true' else False

or

> looking_for_capital = request.POST.get('lookingForCaptail') == 'true'

kziemianek replied:
Yeah i saw this solution, i wonder if there is another...

kziemianek replied:
I mean framework should provide proper deserializer, no manual checks spread across the code

kziemianek replied:
Imma try this tomorrow: https://stackoverflow.com/questions/15476983/deserialize-a-json-string-to-an-object-in-python

mkosowsk replied:
Neat that looks a lot cleaner 👍🏻

kziemianek commented:
Actual implementation fetches disqus thread's data on client side (most recent values, api limits may be exceeded) and there are no likes and posts counts in model so there is no way to perform sort. 

To provide sorting functionality these values should be placed in model but syncing them with disqus may be problematic. 

I can think of two syncing modes:
1. Lazy - values are updated periodically(no risk to exceed api limits, values may be outdated)
2. Eager - values are updated during ideas list/idea fetch(api limits may be exceeded, most recent values, longer request times).

@mbeacom , @owocki does actual implementation or any of syncing modes satisfy You ?

mkosowsk commented:
Looks great!!! :D

I defer to @owocki and @mbeacom but think option number 2 of values are updated during ideas list/idea fetch would be interesting to run with to start. Nice to have those recent values as part of feedback for the user and I don't think we'll be bumping up against API limits to start 🤔

kziemianek commented:
I'll take a look at webhook later, but seems like it's trigerred only by new comments.

kziemianek commented:
Does the counter next to heart icon in top right corner reflect recommends count in disqus thread ?
If not, then I think we can go with webook...

kziemianek commented:
So if we want that kind of sorting we need to somehow sync our model with disqus threads data. 

As @mkosowsk suggested, fetchin disquss threads data right before ideas list fetch might not have a big impact at start (disqus api max page size of 100), but as soon as we hit 200, 300, 400+ ideas on list requests times will get longer. I think we can defer this optimization.

So if we stick with this solution then when request hits ideas/fetch following actions will be taken:

1. disquss threads fetch
2. ideas fetch
3. ideas post_count and like_count update (if changed)
4. ideas fetch with sorting applied

owocki commented:
> So if we want that kind of sorting we need to somehow sync our model with disqus threads data.

hopefully there is an API and we can just refresh the model to the disquis data on the serverside and not have to do it everytime a user hits the page

kziemianek commented:
syncing will occour every 10 minutes

owocki commented:
very excited for this.. let me know when ready to test

kziemianek commented:
@owocki  :checkered_flag: ready to test

kziemianek commented:
@owocki hold on, I'll add i18n.

kziemianek commented:
@owocki finally :checkered_flag:

kziemianek commented on app/retail/models.py:
```diff
@@ -0,0 +1,81 @@
+    """Define the structure of idea"""
+    full_name = models.CharField(max_length=255)
+    email = models.EmailField(max_length=255)
+    github_username = models.CharField(max_length=255)
```
I think it might be better to reference profile here in order to remove redundancy but I see avatar_url spread across models. What do You think? Maybe refactor in another PR ?

mbeacom replied:
@kziemianek Ideally, if we could reference `profile` and add a related name for reverse as: `ideas`, that would be 👍

mbeacom commented on app/app/urls.py:
```diff
@@ -118,6 +118,12 @@
     url(r'^help/portal?', retail.views.portal, name='portal'),
     url(r'^help/pilot?', retail.views.help_pilot, name='help_pilot'),
     url(r'^help/?', retail.views.help, name='help'),
+    path('ideas/new', retail.views.new_idea, name='new_idea'),
```
👍 for new `path`

mbeacom commented on app/dashboard/migrations/0051_merge_20180404_1535.py:
```diff
@@ -0,0 +1,14 @@
+# Generated by Django 2.0.3 on 2018-04-04 22:35
```
Please delete all three of these migration files and rerun: `make migrations`

mbeacom commented on app/retail/management/commands/sync_ideas.py:
```diff
@@ -0,0 +1,69 @@
+
+logging.getLogger("requests").setLevel(logging.WARNING)
+logging.getLogger("urllib3").setLevel(logging.WARNING)
+class Command(BaseCommand):
```
Two new lines between `logging.getLogger` and `class Command` - Also, `logger` doesn't appear to be used to pass logging anywhere.

mbeacom commented on app/retail/migrations/0004_idea_trending_score.py:
```diff
@@ -0,0 +1,18 @@
+# Generated by Django 2.0.3 on 2018-04-04 11:40
```
Delete all of the `retail` migrations and re-run `make migrations`

mbeacom commented on app/retail/models.py:
```diff
@@ -0,0 +1,81 @@
+
+
+class Idea(SuperModel):
+    """Define the structure of idea"""
```
New line between docstring and attrs.

mbeacom commented on app/retail/models.py:
```diff
@@ -0,0 +1,81 @@
+
+    @property
+    def thread_ident(self):
+        return 'idea-'+str(self.id)
```
Can we switch this to an f-string? `return f'idea-{self.id}'

mbeacom commented on app/retail/templates/idea.html:
```diff
@@ -0,0 +1,42 @@
+{% comment %} Copyright (C) 2017 Gitcoin Core This program is free software: you can redistribute it and/or modify it under
```
😂 I've seen this happen in VSCode. 
Can we drop the`{% comment %}` block back down the way it normally is?

kziemianek replied:
yes, i hit format all the time

mbeacom commented on app/retail/templates/idea.html:
```diff
@@ -0,0 +1,42 @@
+    </div>
+  </div>
+
+  {% include 'shared/bottom_notification.html' %} {% include 'shared/analytics.html' %} {% include 'shared/footer_scripts.html'%} 
```
Can we place each include on an individual line for readability?

mbeacom commented on app/retail/templates/idea_template.html:
```diff
@@ -0,0 +1,57 @@
+{% load i18n static %}
+<script id="idea_template" type="text/x-jsrender">
```
Can we fix the indentation to 2-spaces in this file?

kziemianek replied:
yes, format doesn't work here :rofl:

mbeacom commented on app/retail/templates/ideas_list.html:
```diff
@@ -0,0 +1,63 @@
+{% comment %} Copyright (C) 2017 Gitcoin Core This program is free software: you can redistribute it and/or modify it under
```
Can we resolve the comment block here and move includes and what not to their own lines?

mbeacom commented on app/retail/templates/new_idea.html:
```diff
@@ -0,0 +1,120 @@
+{% comment %} Copyright (C) 2017 Gitcoin Core This program is free software: you can redistribute it and/or modify it under
```
Comment block and individual lines

mbeacom commented on app/retail/views.py:
```diff
@@ -461,6 +468,64 @@ def browser_extension_chrome(request):
 def browser_extension_firefox(request):
     return redirect('https://addons.mozilla.org/en-US/firefox/addon/gitcoin/')
 
```
Can you place two new lines between module level methods?

mbeacom commented on app/retail/templates/idea.html:
```diff
@@ -0,0 +1,49 @@
+  General Public License for more details. You should have received a copy of the GNU Affero General Public License along with
+  this program. If not, see <http://www.gnu.org/licenses/>.
+{% endcomment %} 
+{% load static %}
```
Duplicate load `static`

mbeacom commented on app/retail/templates/ideas_list.html:
```diff
@@ -0,0 +1,70 @@
+  General Public License for more details. You should have received a copy of the GNU Affero General Public License along with
+  this program. If not, see <http://www.gnu.org/licenses/>.
+{% endcomment %}
+{% load static %}
```
Duplicate load `static`

mbeacom commented on app/retail/templates/new_idea.html:
```diff
@@ -0,0 +1,122 @@
+  General Public License for more details. You should have received a copy of the GNU Affero General Public License along with
+  this program. If not, see <http://www.gnu.org/licenses/>.
+{% endcomment %}
+{% load static %}
```
Duplicate load `static`

mbeacom commented on app/retail/views.py:
```diff
@@ -496,6 +504,70 @@ def browser_extension_firefox(request):
+def create_idea(request):
+    idea = Idea(**json.loads(request.body))
+    if request.user.is_authenticated:
+        idea.profile = Profile.objects.get(pk=request.user.profile.id)
```
You should be able to use `idea.profile = request.user.profile` here.

mbeacom commented on app/retail/views.py:
```diff
@@ -496,6 +504,70 @@ def browser_extension_firefox(request):
+@require_POST
+@csrf_exempt
+def create_idea(request):
+    idea = Idea(**json.loads(request.body))
```
Just a note, we'll ultimately want to migrate this to a form to handle cleaning and validation on the posted values.

kziemianek replied:
i get it!

mbeacom commented on app/retail/views.py:
```diff
@@ -496,6 +504,70 @@ def browser_extension_firefox(request):
+
+@csrf_exempt
+def new_idea(request):
+    return TemplateResponse(request, 'new_idea.html', {})
```
We should use the same view for `new_idea` and `create_idea`, handling create on POST, and render page on GET.

kziemianek replied:
ok, i get it!

kziemianek commented:
Thanks for your review. I'll post changes in a couple of hours

kziemianek commented on app/retail/views.py:
```diff
@@ -538,6 +544,78 @@ def browser_extension_firefox(request):
+
+
+def new_idea(request):
+    if request.POST:
```
if post create idea and redirect to created idea, otherwise show idea entry form.
Is this what You were asking about ?I can also use ModelForm :)

kziemianek commented:
@mbeacom I've changed code according to Your comments. Thanks for review:)

kziemianek commented:
Sounds great. Hit me up if something needs to be done.

mbeacom commented on app/retail/templates/idea_template.html:
```diff
@@ -0,0 +1,55 @@
+<script id="idea_template" type="text/x-jsrender">
```
Missing i18n load here.  Adding now.

mbeacom commented on app/retail/models.py:
```diff
@@ -0,0 +1,83 @@
+    @property
+    def local_avatar_url(self):
+        if self.profile:
+            return self.profile.local_avatar_url
```
Bad ref

mbeacom commented on app/assets/v2/js/pages/idea.js:
```diff
@@ -0,0 +1,31 @@
+      // ajax request to load the disqus javascript
+      $.ajax({
+        type: 'GET',
+        url: 'http://' + forumName + '.disqus.com/embed.js',
```
Can we load this over HTTPS?

mbeacom commented:
This is on https://stage.gitcoin.co

Example:
https://stage.gitcoin.co/ideas/idea/1/show

mbeacom commented:
@kziemianek Made a few minor tweaks.

It appears the username isn'r rendering on the list view:

<img width="1126" alt="screenshot 2018-05-02 13 50 05" src="https://user-images.githubusercontent.com/7315957/39540185-e388a2fe-4e0f-11e8-842e-ecd1347c2570.png">

mbeacom commented on scripts/crontab:
```diff
@@ -43,5 +43,5 @@ PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/us
 1 10 * * 1 cd gitcoin/coin; bash scripts/run_management_command.bash assemble_leaderboards  >> /var/log/gitcoin/assemble_leaderboards.log  2>&1
 10 * * * * cd gitcoin/coin; bash scripts/run_management_command.bash pull_stats  >> /var/log/gitcoin/pull_stats.log  2>&1
 10 3 * * * cd gitcoin/coin; bash scripts/run_management_command.bash pull_github  >> /var/log/gitcoin/pull_github.log  2>&1
+*/10 * * * * cd gitcoin/coin; bash scripts/run_management_command.bash sync_ideas 1  >> /var/log/gitcoin/sync_ideas.log  2>&1
```
`sync_ideas` doesn't take any arguments. This crontab fails.

thelostone-mc commented:
@sethmcleod mind checking this out since it was your design -> you might have a better grasp over it 
https://stage.gitcoin.co/ideas/list

@kziemianek 

- Nothing happens when I click on the heart icon ? (I'm logged in)
- the username doesn't turn up next to the `submitted by` 
- Could you update the styling of the dropdown (checkout the `bounty/new`) The CSS is already there, you just need to update the html
- Also the page is using a mix of Futura & Muli. Could we stick to Muli ? 
- Rather than defining the font-size , to ensure uniformity could we stick to using the classes defined https://github.com/gitcoinco/web/blob/master/app/assets/v2/css/gitcoin.css#L197-L211
Body -> `font-body`, heading -> `font-title` and so on
- Could we add some breathing space between disqus component and ideas (margin top/bottom of 1em ? )
- Is it just me / does it look odd, when the Has & Needs section are empty. Would it make sense to leave the default entry as `-`

<img width="1440" alt="screen shot 2018-05-03 at 12 17 28 am" src="https://user-images.githubusercontent.com/5358146/39546429-02bed6fe-4e72-11e8-9b83-90373231a645.png">

@PixelantDesign your thoughts ?
Anything obvious I'm missing out @mbeacom @SaptakS ?

kziemianek commented:
@owocki  @mbeacom @thelostone-mc few fixes in last commits:

> It appears the username isn't rendering on the list or detail views:

fixed

> Recommended doesn't display updated count on list or details view.

the command should update this value, it gets updated after a while, you want it updated right after hitting hearth ?

> ‘github username’ input — shouldnt this be just the ’github login

changed label to github login

> when i submit and im not logged in

fixed

> when i submit and i am logged in

unfortunetly I cannot reproduce this problem

> javascript injection! bwahahhahaha! 

cool! fixed :)

> the ‘ideas/new’ and the ‘ideas show’ page need links back to the main ideas page i think

there is a link in navbar, should i add next?

> URIs - why ideas/list and not ideas/ ?

changed to ideas/

> URIs — why ideas/idea/2/show and not ideas/<pk>-<slug> ?

changed to ideas/ideas/<pk>... what is slug?

owocki commented:
> changed to ideas/ideas/... what is slug?

can we change to `ideas/<pk>-<slug>` ? i think the double `ideas` is wasteful in the URI

slug == the `slugify` method on this django util https://docs.djangoproject.com/en/2.0/ref/utils/

owocki commented:
sweet looking fwd to testing now

kziemianek commented:
@owocki sorry the url is `ideas/idea/<int:idea_id>`

kziemianek commented:
@thelostone-mc, i made some changes according to Your comments, thanks a lot :)
here is the summary:

> Nothing happens when I click on the heart icon ? (I'm logged in)

i've described it in last comment

> the username doesn't turn up next to the submitted by

i've described it in last comment

> Could you update the styling of the dropdown (checkout the bounty/new) The CSS is already there, you just need to update the html

You mean these dropdowns with search built-in ?

> Also the page is using a mix of Futura & Muli. Could we stick to Muli ?

Changed it to Muli

> Rather than defining the font-size , to ensure uniformity could we stick to using the classes defined 

does this meet your expectations ? 
![changed_fonts](https://user-images.githubusercontent.com/1838338/39552300-e77d7a16-4e68-11e8-8d1c-305da19b0c82.png)


> Could we add some breathing space between disqus component and ideas (margin top/bottom of 1em ? )

definetly yes!


> Is it just me / does it look odd, when the Has & Needs section are empty. Would it make sense to leave the default entry as -

added - for empty has and needs

thelostone-mc commented:
> You mean these dropdowns with search built-in ?

Yup! We'll throw in the dropdown without search later and update this then

@kziemianek looks better ^_^

kziemianek commented:
@thelostone-mc I also updated select to use jquery select2 with hidden search box.

kziemianek commented:
@mbeacom are we waiting for something? have you seen my last comments :) ?

mbeacom commented on app/retail/views.py:
```diff
@@ -542,6 +548,82 @@ def browser_extension_firefox(request):
+        designers_exists = bool(request.POST.get('designers_exists'))
+        customer_exists = bool(request.POST.get('customer_exists'))
+        profile = request.user.profile if request.user.is_authenticated else None
+        idea = Idea(full_name=full_name, email=email, github_username=github_username, summary=summary, more_info=more_info,
```
v2 we should probably move this to a form.

mbeacom commented:
Seeing the same ideas after clicking `Show More`

<img width="1135" alt="screenshot 2018-05-17 09 04 40" src="https://user-images.githubusercontent.com/7315957/40179079-5b9a2e0a-59b1-11e8-9d3b-7e58d409f999.png">

mbeacom commented:
Also, I `recommended` the cooking discussion (of course! 😆) and it appears we'll need to add custom caching rules to the list page.

kziemianek commented:
@mbeacom i bet duplicated idea was dropped to next page after another idea becoming more trending, so it was returned by paginator again. 

As a workaround, I can add check to ensure only one idea with given `id` is displayed in list.
